### DISCUSSION TITLE
Can Ray send debug information from a Vue PWA running in production?

### DISCUSSION THREAD

**Discussion by vesper8** [spatie/ray#347] [Q&A]
I just purchased a lifetime license and I'm sure I'll find a lot of use and value for this so I'm excited to use it. One of the main pain points I am hoping to solve however is being able to send debug information from a Vue PWA running in a phone. Because it can be such a pain getting access to a debug console when debugging something working on a phone.

I was kind of hoping the way that Ray worked from production environments was that it would proxy the debug call to a server which would then sync it with the local Mac app. But that doesn't seem to be how it works.

#1 Is my use-case fulfillable then?

#2 And do you have any plans to make it easier to receive debug information from production environments. I see right now, for example, to get debug info from production laravel environments you have to SSH into the server. But there would be an alternative way right? If all the debug calls were forwarded to a proxy service which the mac app could be connected to.. this would then work without having to SSH into a server.

Many thanks!

**Comment by patinthehat** [comment]
@vesper8 
1) You could use something like ngrok to do this. Are you using the [vue-ray](https://github.com/permafrost-dev/vue-ray) package as well?
2) This may need to be answered by @freekmurze or another Spatie team member

**Reply by vesper8** [reply]
Thanks. I would like to hear from @freekmurze too on this issue. It would be nice to have a guide on how to get this working using ngrok or expose

As for vue-ray.. @patinthehat, oddly I can't get it to work at all yet. I have it working perfectly with laravel-ray but no cigar at all for vue-ray yet. And it just fails silently too

In my main.js I added 

```

import Vue from 'vue';

import RayPlugin from 'vue-ray/vue2';

Vue.use(RayPlugin, {
  interceptErrors: false,
  host: 'localhost',
  port: 23517,
  showComponentEvents: [
    'created',
    'mounted',
  ],
});
```

And then in a component I copied your example component 

```

<template>
  <div class="flex-col border-r border-gray-200 bg-white overflow-y-auto w-100">
    <div class="about">
      <button
        type="button"
        class="tw-px-2 tw-py-1 tw-text-sm tw-font-medium tw-text-gray-500 tw-bg-white tw-border tw-border-gray-300 hover:tw-bg-gray-50 focus:tw-z-10 focus:tw-outline-none focus:tw-ring-1 focus:tw-ring-indigo-500 focus:tw-border-indigo-500"
        @click="sendToRay()"
      >
        send ref to ray
      </button><br>
      <button
        type="button"
        class="tw-px-2 tw-py-1 tw-text-sm tw-font-medium tw-text-gray-500 tw-bg-white tw-border tw-border-gray-300 hover:tw-bg-gray-50 focus:tw-z-10 focus:tw-outline-none focus:tw-ring-1 focus:tw-ring-indigo-500 focus:tw-border-indigo-500"
        @click="increment()"
      >
        increment data var
      </button><br>
    </div>
    <div
      ref="div1"
      class="w-full flex flex-wrap"
    >
      <div
        ref="div1a"
        class="w-4/12 inline-flex"
      >
        {{ one }}
      </div>
      <div
        ref="div1b"
        class="w-4/12 inline-flex"
      >
        {{ two }}
      </div>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      one: 100,
      two: 22,
    };
  },

  created() {
    console.log('ray.created-1');
    this.$ray().data();
    console.log('ray.created-2');
    this.$ray().track('one');
    console.log('ray.created-3');
  },

  methods: {
    sendToRay() {
      this.$ray().ref('div1');
    },

    increment() {
      this.one += 3;
    },
  },
};
</script>
```

The button render and the component works, but it doesn't send anything to ray. And if you look at the created hook, only the first console.log fires, the other two don't.. as if it is failing silently when it hits `this.$ray`

**Reply by vesper8** [reply]
Following up. Got this working with expose very easily. So now I can get debug data from a live laravel app without having to connect to it via SSH so that's great. I imagine this will work with vue as well, but I still can't get it to work at all (even locally) via vue-app

**Reply by patinthehat** [reply]
@vesper8 That's a bug I wasn't aware of.  
You can use the following to get it working until I can publish a fix:
```js
import Vue from 'vue';
import { RayPlugin } from 'vue-ray/vue2';

Vue.use(RayPlugin.RayPlugin, {
  interceptErrors: false,
  host: 'localhost',
  port: 23517,
  showComponentEvents: [
    'created',
    'mounted',
  ],
});
```

Note the `RayPlugin.RayPlugin` - I'll get a fix in the works right away, but that should cover you in the mean time.  My apologies for the inconvenience!

**Reply by vesper8** [reply]
@patinthehat Thanks : ) I appreciate the quick support

Unfortunately your proposed workaround also doesn't work. There is no change at all when I use the above code.

I also noticed if I do a `console.log(this)` I cannot see a $ray object in the console output. I do see a `sendToRay` method. I tried to use this `sendToRay` but that doesn't do anything either

**Reply by patinthehat** [reply]
> Following up. Got this working with expose very easily. So now I can get debug data from a live laravel app without having to connect to it via SSH so that's great. I imagine this will work with vue as well, but I still can't get it to work at all (even locally) via vue-app

If you can provide a small sample app that reproduces your issue, I'll give you a hand getting it to work.  It's probably something specific to your setup as I was able to reproduce the bug and get it working pretty quickly in a vue 2 app created from the vue cli.

**Reply by vesper8** [reply]
I'll get that to you in a minute. I've made a brand new vue2 app and I can see the error message now 

```
[Vue warn]: Error in created hook: "TypeError: this.$ray is not a function"
```

I guess my more complex app is somehow eating up that error message and not displaying it

**Reply by vesper8** [reply]
Here you go https://github.com/vesper8/rayvue2

This is a brand new vanilla vue-cli app

All I did was add one line to https://github.com/vesper8/rayvue2/blob/master/src/main.js

Created https://github.com/vesper8/rayvue2/blob/master/src/plugins/ray.js

And add your example component in https://github.com/vesper8/rayvue2/blob/master/src/views/Home.vue

And it throws the 
```
[Vue warn]: Error in created hook: "TypeError: this.$ray is not a function"
```

What am I doing wrong?

**Reply by patinthehat** [reply]
Everything looks good, just need a minor tweak to your `plugins/ray.js` file:

```js
import Vue from 'vue';

import { RayPlugin } from 'vue-ray/vue2'; // added braces

Vue.use(RayPlugin.default, {  // added .default
  interceptErrors: false,
  host: 'localhost',
  port: 23517,
  showComponentEvents: [
    'created',
    'mounted',
  ],
});
```
Making these changes worked perfectly for me with the demo project you provided.

**Reply by vesper8** [reply]
Fantastic!! Adding the braces which I missed above, and adding the .default also made it work on my existing complex project : )

And it's also working perfectly through expose so I can send debug info from production!

Thanks so much for the quick support, appreciate it!

**Reply by patinthehat** [reply]
Glad to hear it! Happy to help :+1: 

Stay tuned for a fix to resolve the actual issue.
